In [2]:
import pandas as pd

In [3]:
url = "https://raw.githubusercontent.com/pointOfive/STA130_F23/main/Data/amazonbooks.csv"
df = pd.read_csv(url, encoding='latin1')
df.isna().sum()

Title            0
Author           1
List Price       1
Amazon Price     0
Hard_or_Paper    0
NumPages         2
Publisher        1
Pub year         1
ISBN-10          0
Height           4
Width            5
Thick            1
Weight_oz        9
dtype: int64

In [4]:
rows, columns = df.shape
print(f"The dataset has {rows} rows and {columns} columns.")

The dataset has 325 rows and 13 columns.


Observations are each seperate piece data entered into the data frame. While variables are the different categories of data that each observation has.

In [5]:
df.describe()

List Price  Amazon Price    NumPages     Pub year      Height  \
count  324.000000    325.000000  323.000000   324.000000  321.000000   
mean    18.579753     13.333846  335.857585  2002.206790    8.163240   
std     14.252829     13.727679  161.984389    10.629002    0.918739   
min      1.500000      0.770000   24.000000  1936.000000    5.100000   
25%     13.950000      8.600000  208.000000  1998.000000    7.900000   
50%     15.000000     10.200000  320.000000  2005.000000    8.100000   
75%     19.950000     13.130000  416.000000  2010.000000    8.500000   
max    139.950000    139.950000  896.000000  2011.000000   12.100000   

            Width       Thick   Weight_oz  
count  320.000000  324.000000  316.000000  
mean     5.585000    0.907716   12.487975  
std      0.874057    0.368625    6.644648  
min      4.100000    0.100000    1.200000  
25%      5.200000    0.600000    7.800000  
50%      5.400000    0.900000   11.200000  
75%      5.900000    1.100000   16.000000  
max      9.500000    2.100000   35.200000

In [6]:
df['NumPages'].value_counts()

NumPages
352.0    15
384.0    15
288.0    14
368.0    13
336.0    13
         ..
298.0     1
92.0      1
112.0     1
607.0     1
344.0     1
Name: count, Length: 110, dtype: int64

In [7]:
df.shape

(325, 13)

In [31]:
url2 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

In [32]:
df2 = pd.read_csv(url2, encoding='latin1')

In [33]:
df2.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [34]:
df2.shape

(891, 15)

df.shape will give the amount of all both rows and columns regardless of whether or not they are missing.
df.describe(), will give stats based on the numerical data.
Shape shows the size, the amount of columns and rows a dataset has while describe() shows summaries of the dataFrame.

Attributes will give information on the data itself while methods will mutate or act on an object

https://chatgpt.com/share/66e3a1d0-0c50-8006-bc60-d9725767a238

Post Lecture //

Count gives the amount of cells/data points that are not empty or NA. Mean is the average of the numerical values in each of the valid cells. std is standard deviation which is how spread out the data is. 25%, 50%, 75% are all quartiles of the data. Min and max are the smallest and largest datapoints in the dataframe.

7. 
7.1. dropna() would be preferred over del df['col'] when the columns are more important for the interpretation of the data rather than deleting whole columns that may contain critical information because a few rows were missing data in those columns. If the columns have a lot fo useful data and only a few rows are missing data then it would be better to use dropna()                             7.2 If the column doesn't have useful data and is kind of irrelevant to intereptting the data as well as missing a high amount of data using del df['col'] would be preferred.

7.3 Using them together could help create a dataframe with the most useful parts of the original dataframe, if we use del df['col'] first we can remove irrelvant columns, then using dropna() would only affect the more important datasets and columns. This creates the most optimal data.

In [35]:
df2.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [37]:
df2.shape

(891, 15)

In [38]:
del df2['deck']

In [39]:
del df2['age']

In [43]:
df2.dropna()

survived  pclass     sex  sibsp  parch     fare embarked   class    who  \
0           0       3    male      1      0   7.2500        S   Third    man   
1           1       1  female      1      0  71.2833        C   First  woman   
2           1       3  female      0      0   7.9250        S   Third  woman   
3           1       1  female      1      0  53.1000        S   First  woman   
4           0       3    male      0      0   8.0500        S   Third    man   
..        ...     ...     ...    ...    ...      ...      ...     ...    ...   
886         0       2    male      0      0  13.0000        S  Second    man   
887         1       1  female      0      0  30.0000        S   First  woman   
888         0       3  female      1      2  23.4500        S   Third  woman   
889         1       1    male      0      0  30.0000        C   First    man   
890         0       3    male      0      0   7.7500        Q   Third    man   

     adult_male  embark_town alive  alone  
0          True  Southampton    no  False  
1         False    Cherbourg   yes  False  
2         False  Southampton   yes   True  
3         False  Southampton   yes  False  
4          True  Southampton    no   True  
..          ...          ...   ...    ...  
886        True  Southampton    no   True  
887       False  Southampton   yes   True  
888       False  Southampton    no  False  
889        True    Cherbourg   yes   True  
890        True   Queenstown    no   True  

[889 rows x 13 columns]

In [45]:
df2.dropna(axis=1)

survived  pclass     sex  sibsp  parch     fare   class    who  \
0           0       3    male      1      0   7.2500   Third    man   
1           1       1  female      1      0  71.2833   First  woman   
2           1       3  female      0      0   7.9250   Third  woman   
3           1       1  female      1      0  53.1000   First  woman   
4           0       3    male      0      0   8.0500   Third    man   
..        ...     ...     ...    ...    ...      ...     ...    ...   
886         0       2    male      0      0  13.0000  Second    man   
887         1       1  female      0      0  30.0000   First  woman   
888         0       3  female      1      2  23.4500   Third  woman   
889         1       1    male      0      0  30.0000   First    man   
890         0       3    male      0      0   7.7500   Third    man   

     adult_male alive  alone  
0          True    no  False  
1         False   yes  False  
2         False   yes   True  
3         False   yes  False  
4          True    no   True  
..          ...   ...    ...  
886        True    no   True  
887       False   yes   True  
888       False    no  False  
889        True   yes   True  
890        True    no   True  

[891 rows x 11 columns]

In [51]:
df2.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [47]:
df2.shape

(891, 13)

First I removed the irrelvant columns, the ones with a huge proportion of the data missing then I removed remaining missing data rows

https://chatgpt.com/share/66e3b37d-bd80-8006-9a56-550b33f4dbdd


8.1 df.groupby("col1")["col2"].describe() gives a summary of min, max, the quartiles, count and mean for each column listed.

In [52]:
df.groupby("Height")["Width"].describe()

count      mean       std  min    25%   50%    75%  max
Height                                                         
5.1       1.0  8.500000       NaN  8.5  8.500  8.50  8.500  8.5
5.2       2.0  7.600000  0.565685  7.2  7.400  7.60  7.800  8.0
5.5       2.0  8.200000  0.000000  8.2  8.200  8.20  8.200  8.2
5.6       1.0  8.200000       NaN  8.2  8.200  8.20  8.200  8.2
5.7       1.0  4.300000       NaN  4.3  4.300  4.30  4.300  4.3
5.8       2.0  4.350000  0.070711  4.3  4.325  4.35  4.375  4.4
6.3       2.0  4.350000  0.212132  4.2  4.275  4.35  4.425  4.5
6.6       3.0  4.500000  0.608276  4.1  4.150  4.20  4.700  5.2
6.7       5.0  4.140000  0.054772  4.1  4.100  4.10  4.200  4.2
6.8       3.0  4.233333  0.152753  4.1  4.150  4.20  4.300  4.4
6.9       4.0  4.950000  0.967815  4.1  4.175  4.80  5.575  6.1
7.0       1.0  5.300000       NaN  5.3  5.300  5.30  5.300  5.3
7.1       4.0  4.900000  0.432049  4.3  4.750  5.00  5.150  5.3
7.2       2.0  4.900000  0.989949  4.2  4.550  4.90  5.250  5.6
7.3       2.0  5.150000  0.212132  5.0  5.075  5.15  5.225  5.3
7.4       2.0  5.600000  0.565685  5.2  5.400  5.60  5.800  6.0
7.5       7.0  4.814286  0.693851  4.2  4.300  4.30  5.450  5.7
7.6       7.0  5.157143  0.127242  4.9  5.150  5.20  5.200  5.3
7.7       5.0  5.080000  0.083666  5.0  5.000  5.10  5.100  5.2
7.8      20.0  5.145000  0.274293  4.2  5.100  5.15  5.300  5.5
7.9      33.0  5.221212  0.136376  5.0  5.100  5.20  5.300  5.6
8.0      44.0  5.252273  0.214030  4.2  5.200  5.20  5.400  5.7
8.1      19.0  5.378947  0.165257  5.2  5.250  5.40  5.450  5.8
8.2      27.0  5.422222  0.217208  5.2  5.300  5.40  5.500  6.2
8.3      21.0  5.480952  0.390665  5.0  5.200  5.40  5.700  6.4
8.4      13.0  5.446154  0.171345  5.2  5.400  5.40  5.500  5.8
8.5      10.0  5.790000  1.267061  5.0  5.200  5.45  5.725  9.3
8.6       2.0  5.600000  0.141421  5.5  5.550  5.60  5.650  5.7
8.7       3.0  5.900000  0.400000  5.5  5.700  5.90  6.100  6.3
8.8       3.0  6.266667  0.461880  6.0  6.000  6.00  6.400  6.8
8.9      10.0  5.990000  0.137032  5.8  5.900  6.00  6.000  6.3
9.0       7.0  6.385714  0.511301  6.0  6.000  6.10  6.700  7.2
9.1      14.0  6.271429  0.489000  5.6  6.100  6.10  6.200  7.4
9.2       3.0  6.266667  0.351188  5.9  6.100  6.30  6.450  6.6
9.3       5.0  6.560000  0.207364  6.3  6.400  6.60  6.700  6.8
9.4       6.0  6.550000  0.450555  6.2  6.300  6.35  6.625  7.4
9.5       5.0  6.600000  0.504975  6.0  6.500  6.50  6.600  7.4
9.6       3.0  6.400000  0.100000  6.3  6.350  6.40  6.450  6.5
10.3      1.0  7.900000       NaN  7.9  7.900  7.90  7.900  7.9
10.6      1.0  8.200000       NaN  8.2  8.200  8.20  8.200  8.2
10.7      1.0  8.900000       NaN  8.9  8.900  8.90  8.900  8.9
10.8      1.0  8.400000       NaN  8.4  8.400  8.40  8.400  8.4
11.0      1.0  9.000000       NaN  9.0  9.000  9.00  9.000  9.0
12.1      1.0  9.200000       NaN  9.2  9.200  9.20  9.200  9.2

In [53]:
df.describe()

List Price  Amazon Price    NumPages     Pub year      Height  \
count  310.000000    310.000000  310.000000   310.000000  310.000000   
mean    17.934516     12.587645  330.577419  2002.035484    8.130968   
std     13.287255     11.578859  159.017713    10.732796    0.868884   
min      1.500000      0.770000   24.000000  1936.000000    5.100000   
25%     13.425000      8.537500  208.000000  1998.000000    7.900000   
50%     15.000000     10.200000  320.000000  2005.000000    8.100000   
75%     18.000000     12.240000  403.750000  2010.000000    8.400000   
max    139.950000    139.950000  896.000000  2011.000000   12.100000   

            Width       Thick   Weight_oz  
count  310.000000  310.000000  310.000000  
mean     5.553548    0.890000   12.427742  
std      0.845416    0.355534    6.622459  
min      4.100000    0.100000    1.200000  
25%      5.200000    0.600000    7.800000  
50%      5.300000    0.800000   11.200000  
75%      5.800000    1.100000   16.000000  
max      9.300000    2.000000   35.200000

8.2 df.describe() gives a statistical overview of the whole dataframe while df.groupby[][].describe() shows the data grouped by the columns included, also provides information on how data varies in each group included.

8.3 for me they were both equaly as fast, all I needed to do on good was click a drop down and it gave me the answer whle chatgpt also gave me the answer quickly, it just generated the code for importing pandas. I think chatgpt may be easier to use if it is already open, it also identifies the answer and problem.

8.3B I mistyped titanics and chatgpt fixed the typo. Google identified the problem that there was no file named that in the directory.

8.3C I got a NameError both chatgpt and google gave me a quick and easy answer

8.3D For this error chatgpt was quicker at finding an answer, google I had to visit a seperate website to find my asnwer

8.3E This was an attributeError, chatgpt fixed it very fast and identified that group_by was instead supposed to be groupby. However google also had the answer in a drop down saying I tried to access an attribute that does not exsist.

8.3F I got a keyError, when searching keyError in google, google said that I was trying to access a page or dictionary that did not exsist. Same for chatgpt, but chatGPT gives examples and solutions to those examples.

8.3G This gave another nameError Both chatgpt and google were fast again, google had it in a drop down saying tried to access a variable or function that has not been defined. ChatGPT can take the whole error message and gave me a solution. I probably would prefer chatGPT over google since it identifies and offers solutions, but google is also good for just quick and easy searches that are not overly complicated.

https://chatgpt.com/share/66e3c358-450c-8006-84aa-a4cc16728c8b

no